src/Orleans/Providers/DefaultServiceProvider.cs

```diff
@@ -8,8 +10,47 @@ namespace Orleans.Providers
     /// </summary>
     public class DefaultServiceProvider : IServiceProvider
     {
+        private readonly IDictionary<Type, object> singletonsByType;
+        private readonly IDictionary<Type, object> singletonsByInterface;
+
+        public DefaultServiceProvider()
+            : this(new Dictionary<Type,object>() )
+        {
+        }
+
+        public DefaultServiceProvider(IDictionary<Type, object> singletons)
+        {
+            if (singletons == null)
+            {
+                throw new ArgumentNullException("singletons");
+            }
+
+            // if singleton type is interface, add to list of singleons obtainable via interface.
+            singletonsByInterface = singletons.Where(kvp => kvp.Key.IsInterface)
+                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+            // if type is not interface, as to list of singletons accessable by type
+            singletonsByType = singletons.Where(kvp => !kvp.Key.IsInterface)
+                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+
+            // regardless of type mapping provided, always map singleton to is actuall type
```

Comment:
Comment typo:
// Regardless of type mapping provided, always map singleton to it's actual type.